bin/generate.dart

```diff
@@ -353,23 +386,87 @@ class DiagramGenerator {
   }
 }
 
+Future<Map<String, Map<String, String>>> listAvailableDevices() async {
+  final ProcessRunnerResult result = await ProcessRunner().runProcess(
+    <String>['flutter', 'devices', '--machine'],
+    printOutput: false,
+  );
+
+  final Map<String, Map<String, String>> devices =
+      <String, Map<String, String>>{};
+  final List<dynamic> devicesJson = jsonDecode(result.stdout) as List<dynamic>;
+  for (final Map<String, dynamic> entry
+      in devicesJson.cast<Map<String, dynamic>>()) {
+    devices[(entry['name'] as String).toLowerCase()] = <String, String>{
+      'id': entry['id'] as String,
+      'targetPlatform': entry['targetPlatform'] as String
+    };
+  }
+  return devices;
+}
+
+String getDeviceList(Map<String, Map<String, String>> devices) {
+  final List<String> output = <String>[];
+  for (final String key in devices.keys) {
+    final String id = devices[key]!['id']!;
+    final String platform = devices[key]!['targetPlatform']!;
+    output.add('$key : $id ${id != platform ? ' ($platform)' : ''}');
+  }
+  return output.join('\n');
+}
+
+// This will default to macOS on macOS, linux on linux, and the first attached
+// Android device on Windows.
+String getDefaultDevice(Map<String, Map<String, String>> devices) {
+  if (devices.isEmpty) {
+    return '';
+  }
+  final String platform = Platform.operatingSystem.toLowerCase();
+  if (devices.keys.contains(platform)) {
+    if (platform != 'windows') {
+      // We don't yet support Windows as a device, since the process requires
+      // some Unix tools still (ffmpeg, optipng, etc.).
```

Comment:
i don't understand the situation on windows. if it needs tools, shouldn't we just not do anything? what does it mean to default to anything on windows?